src/geode/model/helpers/model_concatener.cpp

```diff
@@ -0,0 +1,255 @@
+        geode::BRepBuilder& builder,
+        const geode::BRep& other_model,
+        const geode::ModelCopyMapping& mapping )
+    {
+        for( const auto& block : other_model.blocks() )
+        {
+            const auto& block_copy = model.block(
+                mapping.at( geode::Block3D::component_type_static() )
+                    .in2out( block.id() ) );
+            for( const auto& surface : other_model.boundaries( block ) )
+            {
+                const auto& surface_copy = model.surface(
+                    mapping.at( geode::Surface3D::component_type_static() )
+                        .in2out( surface.id() ) );
+                builder.add_surface_block_boundary_relationship(
+                    surface_copy, block_copy );
+            }
+            for( const auto& surface : other_model.internal_surfaces( block ) )
+            {
+                const auto& surface_copy = model.surface(
+                    mapping.at( geode::Surface3D::component_type_static() )
+                        .in2out( surface.id() ) );
+                builder.add_surface_block_internal_relationship(
+                    surface_copy, block_copy );
+            }
+            for( const auto& line : other_model.internal_lines( block ) )
+            {
+                const auto& line_copy = model.line(
+                    mapping.at( geode::Line3D::component_type_static() )
+                        .in2out( line.id() ) );
+                builder.add_line_block_internal_relationship(
+                    line_copy, block_copy );
+            }
+            for( const auto& corner : other_model.internal_corners( block ) )
+            {
+                const auto& corner_copy = model.corner(
+                    mapping.at( geode::Corner3D::component_type_static() )
+                        .in2out( corner.id() ) );
+                builder.add_corner_block_internal_relationship(
+                    corner_copy, block_copy );
+            }
+        }
+    }
+} // namespace
+
+namespace geode
+{
+    template < typename Model >
+    class ModelConcatener< Model >::Impl
+    {
+        using ModelBuilder = typename Model::Builder;
+        static constexpr auto dimension = Model::dim;
+
+    public:
+        Impl( Model& model ) : model_( model ), builder_{ model } {}
+
+        ModelCopyMapping concatenate( const Model& other_model )
+        {
+            const auto mapping = builder_.copy_components( other_model );
+            builder_.copy_component_geometry( mapping, other_model );
+            copy_relationships( other_model, mapping );
+            return mapping;
+        }
+
+    private:
+        void copy_relationships(
+            const Model& other_model, const ModelCopyMapping& mapping );
+
+        void copy_line_relationships(
+            const Model& other_model, const ModelCopyMapping& mapping )
+        {
+            for( const auto& line : other_model.lines() )
```

Comment:
You could get the line_mapping and corner_mapping once a the beginning of the method to avoid having the mapping.at() each time

This could be apply for every methods